python/ray/serve/_private/router.py

```diff
@@ -83,62 +125,194 @@ def update_running_replicas(self, running_replicas: List[RunningReplicaInfo]):
             )
+            yield [self._replicas[chosen_id] for chosen_id in chosen_ids]
 
-        return replica.actor_handle.handle_request_streaming.options(
-            num_returns="streaming"
-        ).remote(
-            pickle.dumps(query.metadata),
-            *query.args,
-            **query.kwargs,
+            await asyncio.sleep(self.backoff_sequence_s[backoff_index])
+            backoff_index = min(backoff_index + 1, len(self.backoff_sequence_s) - 1)
+
+    async def select_replica(
+        self, replicas: List[ReplicaWrapper]
+    ) -> Optional[ReplicaWrapper]:
+        tasks = [r.get_queue_len() for r in replicas]
+        done, pending = await asyncio.wait(
+            tasks, timeout=0.1, return_when=asyncio.ALL_COMPLETED
         )
 
-    def update_running_replicas(self, running_replicas: List[RunningReplicaInfo]):
-        new_replica_ids = {r.replica_tag for r in running_replicas}
-        if new_replica_ids != self._replica_id_set:
-            self._replica_id_set = new_replica_ids
-            logger.info(
-                "Got updated replicas for deployment "
-                f"{self._deployment_name}: {new_replica_ids}.",
-                extra={"log_to_stderr": False},
+        chosen_replica_id = None
+        lowest_queue_len = math.inf
+        for task in done:
+            if task.exception() is not None:
+                print(f"Exception: {task.exception()}")
+            else:
+                replica_id, queue_len, accepted = task.result()
+                print(task.result())
+                # print(replica_id, "queue_len:", queue_len)
+                if accepted and queue_len < lowest_queue_len:
+                    chosen_replica_id = replica_id
+                    lowest_queue_len = queue_len
+
+        for task in pending:
+            task.cancel()
+
+        if chosen_replica_id is None:
+            print("No available replicas.")
+            return None
+
+        return self._replicas[chosen_replica_id]
+
+    def schedule_replica(self, replica):
+        while len(self._pending_assignment_futures) > 0:
+            fut = self._pending_assignment_futures.pop(0)
+            if not fut.done():
+                fut.set_result(replica)
+                break
+
+        return len(self._scheduling_tasks) <= self.target_scheduling_tasks
+
+    async def try_schedule(self):
+        curr_task = asyncio.current_task().get_name()
+        print(curr_task, "try_schedule entered")
+        keep_scheduling = True
+        while keep_scheduling:
+            async for candidates in self.choose_two_gen():
+                print("GOT CANDIDATES", candidates)
+                replica = await self.select_replica(candidates)
+                print("GOT REPLICA", replica)
+                if replica is not None:
+                    keep_scheduling = self.schedule_replica(replica)
+                    print(curr_task, "Got replica! Keep scheduling?", keep_scheduling)
```

Comment:
If you `keep_scheduling`, you should reset the backoff inside the `choose_two_gen` generator. Otherwise, it'll keep using the latest backoff instead of restarting from the shortest backoff.